client/client.go

```diff
@@ -0,0 +1,157 @@
+// see the NOTICE file and/or the repository at
+// https://github.com/direct-state-transfer/dst-go
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/channel/persistence/keyvalue"
+	"perun.network/go-perun/client"
+	"perun.network/go-perun/peer"
+	"perun.network/go-perun/pkg/sortedkv/leveldb"
+
+	"github.com/direct-state-transfer/dst-go"
+	"github.com/direct-state-transfer/dst-go/blockchain/ethereum"
+)
+
+// Client is a wrapper type around the state channel client implementation from go-perun.
+// It also manages the lifecycle of listener and handler go-routines.
+type Client struct {
+	*client.Client
+
+	wg *sync.WaitGroup
+}
+
+// NewEthereumPaymentClient initializes a two party, ethereum payment channel client for the given user.
+// It establishes a connection to the blockchain and verifies the integrity of contracts at the given address.
+// It uses the comm backend to initialize adapters for off-chain communication network.
+func NewEthereumPaymentClient(cfg Config, user dst.User, comm dst.CommBackend) (*Client, error) {
+	dialer, listener, err := initComm(comm, user.CommAddr)
+	if err != nil {
+		return nil, err
+	}
+	funder, adjudicator, err := connectToChain(cfg.Chain, user.OnChain)
+	if err != nil {
+		return nil, err
+	}
+	// Only off-chain account is unlocked. Accounts for the participant addresses
+	// will be unlocked by the client when required.
+	offChainAcc, err := user.OffChain.Wallet.Unlock(user.OffChain.Addr)
+	if err != nil {
+		return nil, errors.WithMessage(err, "off-chain address")
+	}
+
+	c := client.New(offChainAcc, dialer, funder, adjudicator, user.OffChain.Wallet)
+	if err := loadPersister(c, cfg.DatabaseDir, cfg.PeerReconnTimeout); err != nil {
+		return nil, err
+	}
+	client := &Client{
+		Client: c,
+		wg:     &sync.WaitGroup{},
+	}
+
+	client.runAsGoRoutine(func() { client.Handle(&proposalHandler{}, &updateHandler{}) })
+	client.runAsGoRoutine(func() { client.Listen(listener) })
+
+	return client, nil
+}
+
+// Close closes the client and waits until the listener and handler go routines return.
+func (c *Client) Close() error {
+	err := c.Client.Close()
+	c.wg.Wait()
+	return errors.Wrap(err, "closing channel client")
+}
+
+func initComm(comm dst.CommBackend, listenAddr string) (peer.Dialer, peer.Listener, error) {
+	if comm == nil {
```

Comment:
Only `comm` it checked because it it is nil, its usage in the following lines will cause the program to panic and shutdown abruptly. 

I do it here, instead of the `NewEthereumPaymentClient` because the usage of the interface is here and it is better to have the nil validation closer to the usage.

Reply:
I would also not use panic, because it will cause the node to stop abruptly, it is better to have a gracious shutdown in case of errors.

Reply:
When giving it another thought, I think, input validation (for nil value of interfaces) is completely not required in this package. Because values for these interfaces will be initialized and injected by the node package. 

if all the validation with respect to interfaces are implemented in the node package, then it would be safe to assume the interface arguments in the functions off all component packages will be supplied with non nil values. As a consequence, the node will be panic due to a nil interface in the component package.